src/monodroid/jni/embedded-assemblies.cc

```diff
@@ -52,13 +77,80 @@ const char *EmbeddedAssemblies::suffixes[] = {
 	".exe",
 };
 
+EmbeddedAssemblies::EmbeddedAssemblies ()
+	: system_page_size (monodroid_getpagesize())
+{}
+
 void EmbeddedAssemblies::set_assemblies_prefix (const char *prefix)
 {
 	if (assemblies_prefix_override != nullptr)
 		delete[] assemblies_prefix_override;
 	assemblies_prefix_override = prefix != nullptr ? utils.strdup_new (prefix) : nullptr;
 }
 
+template<typename T>
+T
+EmbeddedAssemblies::get_mmap_file_data (XamarinBundledAssembly& xba)
+{
+	if (xba.mmap_area == nullptr)
+		mmap_apk_file (xba);
+	return static_cast<T>(xba.mmap_file_data);
+}
+
+inline void
+EmbeddedAssemblies::load_assembly_debug_info_from_bundles (const char *aname)
+{
+	if (!register_debug_symbols || !bundled_assemblies_have_debug_info || aname == nullptr)
+		return;
+
+	size_t aname_len = strlen (aname);
+	for (size_t i = 0; i < bundled_assemblies_count; i++) {
+		XamarinBundledAssembly &entry = bundled_assemblies[i];
+		if (entry.type != FileType::DebugInfo)
+			continue;
+
+		// We know the entry has one of the debug info extensions, so we can take this shortcut to
+		// avoid having to allocate memory and use string comparison in order to find a match for
+		// the `aname` assembly
+		if (strncmp (entry.name, aname, aname_len) != 0)
+			continue;
+
+		size_t ext_len = strlen (entry.name) - aname_len;
+		if (ext_len != 4 || ext_len != 8) { // 'Assembly.pdb' || 'Assembly.{exe,dll}.mdb'
```

Comment:
I believe that this logic is wrong.

I applied this patch:

```diff
diff --git a/src/monodroid/jni/embedded-assemblies.cc b/src/monodroid/jni/embedded-assemblies.cc
index 0577db22..3025efec 100644
--- a/src/monodroid/jni/embedded-assemblies.cc
+++ b/src/monodroid/jni/embedded-assemblies.cc
@@ -100,6 +100,7 @@ EmbeddedAssemblies::get_mmap_file_data (XamarinBundledAssembly& xba)
 inline void
 EmbeddedAssemblies::load_assembly_debug_info_from_bundles (const char *aname)
 {
+       log_warn (LOG_DEFAULT, "# jonp: load_assembly_debug_info_from_bundles: aname=%s", aname);
        if (!register_debug_symbols || !bundled_assemblies_have_debug_info || aname == nullptr)
                return;
 
@@ -112,11 +113,13 @@ EmbeddedAssemblies::load_assembly_debug_info_from_bundles (const char *aname)
                // We know the entry has one of the debug info extensions, so we can take this shortcut to
                // avoid having to allocate memory and use string comparison in order to find a match for
                // the `aname` assembly
+               log_warn (LOG_DEFAULT, "# jonp: load_assembly_debug_info_from_bundles: strncmp(%s, %s, %i)", entry.name, aname, (int) aname_len);
                if (strncmp (entry.name, aname, aname_len) != 0)
                        continue;
 
                size_t ext_len = strlen (entry.name) - aname_len;
                if (ext_len != 4 || ext_len != 8) { // 'Assembly.pdb' || 'Assembly.{exe,dll}.mdb'
+                       log_warn (LOG_DEFAULT, "# jonp: load_assembly_debug_info_from_bundles: extension mismatch");
                        continue;
                }
 
```

`adb logcat` output contains:

```
# jonp: load_assembly_debug_info_from_bundles: aname=UnnamedProject
# jonp: load_assembly_debug_info_from_bundles: strncmp(UnnamedProject.pdb, UnnamedProject, 14)
# jonp: load_assembly_debug_info_from_bundles: extension mismatch
# jonp: load_assembly_debug_info_from_bundles: strncmp(FormsViewGroup.pdb, UnnamedProject, 14)
# jonp: load_assembly_debug_info_from_bundles: strncmp(Xamarin.Forms.Core.pdb, UnnamedProject, 14)
# jonp: load_assembly_debug_info_from_bundles: strncmp(Xamarin.Forms.Platform.Android.pdb, UnnamedProject, 14)
# jonp: load_assembly_debug_info_from_bundles: strncmp(Xamarin.Forms.Xaml.pdb, UnnamedProject, 14)
# jonp: load_assembly_debug_info_from_bundles: strncmp(Mono.Android.pdb, UnnamedProject, 14)
```

In order for `mono_register_symfile_for_assembly()` to be called, we must *not* print `extension mismatch`, yet we do.

One #facepalm later, and it's because of the `||`!  `ext_len` is 4, but it *isn't* 8, and the result of `ext_len != 4 || ext_len != 8` is True, because `ext_len != 8`!

This should be `&&`!

Reply:
Well spotted! The debug files are now registered properly, but the same tests still fail :(